### PR TITLE
Fix: Tooltips can be off screen

### PR DESCRIPTION
**Pull request by nkappler**
Tooltips and the scrollbar inside the hint container can be off screen, which can induce scrollbars on the document body.

image
image
image

Only related bug I could find:
Fix: #5622

### CODE REVIEW

**Comment by nkappler** [comment]
The test seems to fail because of this line:
```javascript
    var winW = window.innerWidth || Math.max(ownerDocument.body.offsetWidth, ownerDocument.documentElement.offsetWidth);
    var winH = window.innerHeight || Math.max(ownerDocument.body.offsetHeight, ownerDocument.documentElement.offsetHeight);
```

Apparently using ownerDocument is not wanted?
I copied this line from show-hint.js so I am not sure why this is a problem.
I would be fine with just using `window.innerWidth`

Maybe one of the reviewers has more knowledge about this?

**Comment by marijnh** [comment]
> Apparently using ownerDocument is not desired?

It's not defined—the linter is catching a reference to a variable that doesn't exist.

**Comment by nkappler** [comment]
> It's not defined—the linter is catching a reference to a variable that doesn't exist.

Got it... would you recommend to add the variable 
```javascript 
var ownerDocument = cm.getInputField().ownerDocument;
```
or rather just use `window.innerWidth`?

**Comment by marijnh** [comment]
I'm fine with just using the window size.

**Comment by nkappler** [comment]
thanks for your help :) 
The fix is now complete in my opinion. Let me know if you have any open questions...

File: addon/hint/show-hint.js [hunk 1]
```diff
@@ -268,8 +268,10 @@
     var winH = parentWindow.innerHeight || Math.max(ownerDocument.body.offsetHeight, ownerDocument.documentElement.offsetHeight);
     container.appendChild(hints);
 
-    var box = completion.options.moveOnOverlap ? hints.getBoundingClientRect() : new DOMRect();
     var scrolls = completion.options.paddingForScrollbar ? hints.scrollHeight > hints.clientHeight + 1 : false;
+    if (scrolls) for (var node = hints.firstChild; node; node = node.nextSibling)
+      node.style.paddingRight = cm.display.nativeBarWidth + "px"
```
Comment (marijnh) [inline addon/hint/show-hint.js:273]: Why did you move this code? This ordering forces another relayout.

Comment (nkappler) [reply addon/hint/show-hint.js:273]: Before my change the code would check if the hints container is off screen. If so, it would move it to be aligned with the edge of the screen. After that, it checks if the hints container has a scrollbar and if so, it applies padding to each hint element which causes the container to grow and be off screen again. (see attached gif)

I just changed the order so that the padding for the scrollbar is applied before it is checked if the container is off screen.

Comment (marijnh) [reply addon/hint/show-hint.js:273]: Yes, but this code has been carefully built to _not_ create unnecessary relayouts, and I don't really want to accept a patch that degrades.

Comment (nkappler) [reply addon/hint/show-hint.js:273]: I don't get why it should create another relayout, it just moves the container once in any case, or am I missing something?

Comment (marijnh) [reply addon/hint/show-hint.js:273]: See for example [https://eloquentjavascript.net/14_dom.html#p_LcMNzqCMUS](this link). Your change causes a write to the DOM to happen between two layout reads, which will force an extra relayout.

Comment (nkappler) [reply addon/hint/show-hint.js]: I see... I actually read your book but didn't remember that section.
Didn't think of that! 

I did some measurements and my change would have added around 10ms of rendering time.
I now have a more elegant solution: If there are scrollbars, the amount of padding added to the items is just added to the `overlapX`value, so no re-layout necessary anymore.
My measurements confirm that it adds no additional rendering time this way:

image
image

File: addon/hint/show-hint.js [hunk 2]
```diff
@@ -293,6 +293,7 @@
       }
     }
     var overlapX = box.right - winW;
+    if (scrolls) overlapX += cm.display.nativeBarWidth;
```
Comment (nkappler) [inline addon/hint/show-hint.js:296]: as mentioned, the amount of padding used in the hint items when there is a scrollbar is also added to the overlapX, moving the hints container just a smidge to the left, so the scrollbar is not off screen...

Comment (marijnh) [reply addon/hint/show-hint.js:296]: Much better, thanks!

Comment (nkappler) [reply addon/hint/show-hint.js:296]: thank YOU for building such awesome tools! :)
